capi/src/inspect.rs

```diff
@@ -267,10 +276,14 @@ pub unsafe extern "C" fn blaze_inspect_syms_elf(
     names: *const *const c_char,
     name_cnt: usize,
 ) -> *const *const blaze_sym_info {
+    if !input_zeroed!(src, blaze_inspect_elf_src) {
+        return ptr::null()
+    }
+
     // SAFETY: The caller ensures that the pointer is valid.
     let inspector = unsafe { &*inspector };
     // SAFETY: The caller ensures that the pointer is valid.
-    let src = Source::Elf(Elf::from(unsafe { &*src }));
+    let src = Source::Elf(Elf::from(unsafe { src.read_unaligned() }));
```

Comment:
read_unaligned() will just read `sizeof(struct blaze_inspect_elf_src)` bytes, right? If yes, then it's problematic, because if user was compiled against older version of the library, they provided not enough bytes, so you'd be reading out of bounds.

On libbpf side I have OPTS_GET() macro, which allows to fetch information field-by-field, using default value, if user didn't provide enough bytes.

An alternative to this is to do what kernel does. Have a local full struct, memset(0) it, copy however many bytes user provided over local struct variable, and then use local variable, knowing that all extra fields are zero. This simplifies code a lot at the price of a tiny mem copy overhead. But you can actually structure code to avoid this overhead if user provided exactly the right amount of bytes (or even more).

Reply:
Yep. I'd say it's fine; once we add additional fields we need to cope with it. This version of the library will never have additional fields, and none will ever have fewer -- so there cannot be any out of bounds reads. Only once we add a field we need to make sure to make it optional, I think.

Reply:
I guess I see no reason to not solve this upfront consistently for all use cases where you add size field, while all this is fresh. Technically, someone can just allocate 8 bytes for size field and pass such "noop" options, which wouldn't be illegal under this sized options contract, even though it's clearly a bit of an abuse.

Anyways, I think it's good to think and handle this upfront once and never have to think about this again, that's my only point.